File: yt/java/ytsaurus-testlib/src/main/java/tech/ytsaurus/testlib/ExceptionUtils.java
package tech.ytsaurus.testlib;

import tech.ytsaurus.lang.NonNullApi;

@NonNullApi
public class ExceptionUtils {
    private ExceptionUtils() {
    }

    public static <T extends Throwable> T getCause(Throwable ex, Class<T> causeClass) {
        for (Throwable cur = ex; cur != null; cur = cur.getCause()) {
            if (causeClass.isInstance(cur)) {
                //noinspection unchecked
                return (T) cur;
            }
        }
        throw new RuntimeException("Exception " + ex + " is not caused by " + causeClass);
    }
}

File: ytsaurus/yt/spark/spark-over-yt/data-source/src/main/scala/tech/ytsaurus/spyt/serialization/ExceptionUtils.scala
package tech.ytsaurus.spyt.serialization

import java.io.{FileNotFoundException, IOException}
import java.util.concurrent.{ExecutionException, TimeoutException}

object ExceptionUtils {
  class RuntimeIoException(cause: Throwable) extends RuntimeException(cause)
  class FileNotFoundIoException(cause: Throwable) extends RuntimeException(cause)
  class InterruptedRuntimeException(cause: Throwable) extends RuntimeException(cause)
  class ExecutionRuntimeException(cause: Throwable) extends RuntimeException(cause)
  class TimeoutRuntimeException(cause: Throwable) extends RuntimeException(cause)

  def translate(e: Exception): RuntimeException = e match {
    case exception: RuntimeException => exception
    case _: FileNotFoundException => new FileNotFoundIoException(e)
    case _: IOException => new RuntimeIoException(e)
    case _: InterruptedException => new InterruptedRuntimeException(e)
    case _: ExecutionException => new ExecutionRuntimeException(e.getCause)
    case _: TimeoutException => new TimeoutRuntimeException(e)
    case _ => new RuntimeException(e)
  }
}
